### PR TITLE
Explicitly add unix defines to clang args

### PR DESCRIPTION
**Pull request by jgravelle-google**
Fixes zlib for wasm backend.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1306,6 +1306,7 @@ def get_bitcode_args(input_files):
           args += shared.EMSDK_CXX_OPTS
         if not shared.Building.can_inline():
           args.append('-fno-inline-functions')
+        args += ['-Dunix', '-D__unix', '-D__unix__']
```
Comment (dschuff) [inline emcc.py]: I think there's already some code in `tools/shared.py` that defines `EMSCRIPTEN` for the wasm backend, this can probably go there.

Comment (jgravelle-google) [inline emcc.py]: That is a much better place for that, thanks.

**Comment by kripken** [comment]
Looks ok to me, but I'd like to understand this more. Why was it needed for the wasm backend but not the asm.js target?

Also cc @juj to make sure he sees this.

**Comment by dschuff** [comment]
The asm.js target already defines these; it just does so directly from inside clang rather than from emscripten. So this change just makes the wasm match asm.js.

**Comment by kripken** [comment]
I see. So should we remove them from the asm.js backend, if there wasn't a reason to add them to the wasm target?

**Comment by jgravelle-google** [comment]
It depends on how closely the asm.js backend should be tied to the emscripten environment.
Paraphrasing the discussion we had in https://reviews.llvm.org/D26431, emscripten is what's providing the unix-y environment and we don't need to proscribe the type of environment for all the users of the wasm backend, in case someone ever wants to write a windows-y environment or something.

**Comment by kripken** [comment]
I see, thanks. I guess we don't need to change anything then, the asm.js backend is pretty tied in to emscripten.

**Comment by juj** [comment]
I think these getting defined on asm.js side is a bit of an oversight, we have never really explored whether `unix`, `__unix` and `__unix__` should always unconditionally exist. Having these always enabled could be risky in some cases, we are now basically emulating to cater to what zlib currently does, out of convenience. There is no good exact spec of how Emscripten/asm.js/wasm should specifically behave when these defines are present, and how to ensure we'd be adhering to this in detail.

My current thinking is that this would be ok, but there could be a scenario where we might want to add flags to avoid these types of defines being added, for users who want to target Emscripten as a "pure" platform that is different from all the other ones they're targeting. We need to be careful that it will be easy to distinguish between native platforms vs Emscripten, without having to add `#if __linux__ && !__EMSCRIPTEN__` types of checks to detect "real" Linuxes, so we should be carefuly not to go all in with emulating Unixness. Perhaps this is a good granularity.

In any case, we need documentation for these preprocessor defines, added that in #4727.

**Pull request merged**